sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt

```diff
@@ -0,0 +1,140 @@
+
+    override fun createAccount(): Result<Boolean> {
+        val userId = syncDeviceIds.userId()
+        val deviceId = syncDeviceIds.deviceId()
+        val deviceName = syncDeviceIds.deviceName()
+
+        val account: AccountKeys = nativeLib.generateAccountKeys(userId = userId)
+        if (account.result != 0L) return Result.Error(code = account.result.toInt(), reason = "Account keys failed")
+
+        val result =
+            syncApi.createAccount(
+                account.userId,
+                account.passwordHash,
+                account.protectedSecretKey,
+                deviceId,
+                deviceName,
+            )
+
+        return when (result) {
+            is Result.Error -> {
+                Timber.i("SYNC signup failed $result")
+                result
+            }
+            is Result.Success -> {
+                syncStore.userId = userId
+                syncStore.deviceId = deviceId
+                syncStore.deviceName = deviceName
+                syncStore.token = result.data.token
+                syncStore.primaryKey = account.primaryKey
+                syncStore.secretKey = account.secretKey
+                Result.Success(true)
+            }
+        }
+    }
+
+    override fun getAccountInfo(): AccountInfo {
+        if (!isSignedIn()) return AccountInfo()
+
+        return AccountInfo(
+            userId = syncStore.userId.orEmpty(),
+            deviceName = syncStore.deviceName.orEmpty(),
+            deviceId = syncStore.deviceId.orEmpty(),
+            isSignedIn = isSignedIn(),
+        )
+    }
+
+    override fun storeRecoveryCode() {
+        val primaryKey = syncStore.primaryKey ?: return
+        val userID = syncStore.userId ?: return
+        val recoveryCodeJson = Adapters.recoveryCodeAdapter.toJson(RecoveryCode(primaryKey, userID))
+
+        Timber.i("SYNC store recoverCode: $recoveryCodeJson")
+        syncStore.recoveryCode = recoveryCodeJson
+    }
+
+    override fun removeAccount() {
+        syncStore.clearAll()
+    }
+
+    private fun isSignedIn() = !syncStore.primaryKey.isNullOrEmpty()
+
+    private class Adapters {
+        companion object {
+            private val moshi = Moshi.Builder().build()
+            val recoveryCodeAdapter: JsonAdapter<RecoveryCode> =
+                moshi.adapter(RecoveryCode::class.java)
+        }
+    }
+}
+
+data class AccountInfo(
+    val userId: String = "",
+    val deviceName: String = "",
+    val deviceId: String = "",
+    val isSignedIn: Boolean = false,
+)
+
+data class RecoveryCode(
+    val primaryKey: String,
+    val userID: String,
+)
+
+sealed class Result<out R> {
+
+    data class Success<out T>(val data: T) : Result<T>()
+    data class Error(val code: Int = -1, val reason: String) : Result<Nothing>()
+
+    override fun toString(): String {
+        return when (this) {
+            is Success<*> -> "Success[data=$data]"
+            is Error -> "Error[exception=$code, $reason]"
+        }
+    }
+}
```

Comment:
Kotlin stdlib already has a `Result` that has more/better functionality. Do we need our own? I would remove it

Reply:
I would like to replace it as well. The problem I have with kotlin `Result` is that `failure` only accepts a `Throwable`. I would like to be able to add any object as possible failure and not only `Throwables`. what do you think?

Reply:
I guess I am not clear why we want to be so specific and propagate the `ErrorResponse` errors. The error string is not localized anyway so we can't really use it.
I wonder if we're over-designing this atm.

No strong opinions

Reply:
I agree with you.

To give more context, this is just the current approach to allow me to show an error as a Toast in the UI. This is just a convenient way to propagate them into the UI. We are not using those errors for other purposes, and it's ok to show them as they come (no localization) from backend or other internal logic.

➡️ In the future, this will change for sure. We will only propagate errors that need to land on the UI (maybe none of them), and we will not use the current entity of `code`:`message`. Maybe at that point will make sense to drop `Result`.